autobahn/asyncio/component.py

```diff
@@ -312,101 +315,118 @@ def start(self, loop=None):
-                        # If this is a "fatal error" that will never work,
-                        # we bail out now
-                        if isinstance(fail.value, ApplicationError):
-                            if fail.value.error in [u'wamp.error.no_such_realm']:
-                                reconnect[0] = False
-                                self.log.error(u"Fatal error, not reconnecting")
-                                txaio.reject(done_f, fail)
-                                return
-
-                            self.log.error(u"{msg}", msg=fail.value.error_message())
-                            return one_reconnect_loop(None)
-
-                        elif isinstance(fail.value, OSError):
-                            # failed to connect entirely, like nobody
-                            # listening etc.
-                            self.log.info(u"Connection failed: {msg}", msg=txaio.failure_message(fail))
-                            return one_reconnect_loop(None)
-
-                        elif _is_ssl_error(fail.value):
-                            # Quoting pyOpenSSL docs: "Whenever
-                            # [SSL.Error] is raised directly, it has a
-                            # list of error messages from the OpenSSL
-                            # error queue, where each item is a tuple
-                            # (lib, function, reason). Here lib, function
-                            # and reason are all strings, describing where
-                            # and what the problem is. See err(3) for more
-                            # information."
-                            self.log.error(u"TLS failure: {reason}", reason=fail.value.args[1])
-                            self.log.error(u"Marking this transport as failed")
-                            transport.failed()
-                        else:
-                            self.log.error(
-                                u'Connection failed: {error}',
-                                error=txaio.failure_message(fail),
-                            )
-                            # some types of errors should probably have
-                            # stacktraces logged immediately at error
-                            # level, e.g. SyntaxError?
-                            self.log.debug(u'{tb}', tb=txaio.failure_format_traceback(fail))
-                            return one_reconnect_loop(None)
-
-                    txaio.add_callbacks(f, session_done, connect_error)
+                def notify_connect_error(fail):
+                    chain_f = txaio.create_future()
+                    handler_f = self.fire('connectfailure', self, fail.value)
+                    txaio.add_callbacks(
+                        handler_f,
+                        lambda _: txaio.reject(chain_f, fail),
+                        lambda _: txaio.reject(chain_f, fail)
+                    )
+                    return chain_f
 
```

Comment:
This is an interesting idea -- a "connectfailure" notification. I think it should be in a separate PR, probably -- partly so we could discuss the idea itself without delaying fixing the re-connect loop for asyncio. Also, if we *do* add a ``connectfailure`` event, the Twisted side should get it too.

What would user-code do with this? Simply feedback to the user?

Reply:
Yeah further user-side logging of transport failures for one.

If you start a component with multiple transports this will allow user-code to "do something" with the exception raised as each failed connect attempt happens.  Without this, if one of your transports is unconnectable but another succeeds and you get a working session, your component.start() error callback won't be called so you'll never know other than by looking in logs one of your transports didn't work.

Also could be handy for one-shot connect attempts with one transport and max_retries =  0.  'connectfailure' handler will fire and you'll know quickly there's a fundamental connection error without having to look at the exception type in the component.start() error callback which could be a session level exception.

Reply:
Okay, I think that makes sense. What do you think @oberstet? (That is, adding a ``connectfailure`` notification that you can listen to on ``Component`` on both twisted +  asyncio)